builder/chroots.go

```diff
@@ -0,0 +1,637 @@
+package builder
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/clearlinux/mixer-tools/helpers"
+	"github.com/go-ini/ini"
+	"github.com/pkg/errors"
+)
+
+// TODO: Move this to the more general configuration handling.
+type buildChrootsConfig struct {
+	// [Server] section.
+	HasServerSection bool
+	DebugInfoBanned  string
+	DebugInfoLib     string
+	DebugInfoSrc     string
+
+	// [swupd] section.
+	UpdateBundle string
+	ContentURL   string
+	VersionURL   string
+	// Format is already in b.Format.
+}
+
+// TODO: Move this to the more general configuration handling.
+func readBuildChrootsConfig(path string) (*buildChrootsConfig, error) {
+	iniFile, err := ini.InsensitiveLoad(path)
+	if err != nil {
+		return nil, err
+	}
+
+	cfg := &buildChrootsConfig{}
+
+	// TODO: Validate early the fields we read.
+	server, err := iniFile.GetSection("Server")
+	if err == nil {
+		cfg.HasServerSection = true
+		cfg.DebugInfoBanned = server.Key("debuginfo_banned").Value()
+		cfg.DebugInfoLib = server.Key("debuginfo_lib").Value()
+		cfg.DebugInfoSrc = server.Key("debuginfo_src").Value()
+	}
+
+	swupd, err := iniFile.GetSection("swupd")
+	if err != nil {
+		return nil, fmt.Errorf("error in configuration file %s: %s", path, err)
+	}
+
+	getKey := func(section *ini.Section, name string) (string, error) {
+		key, kerr := section.GetKey(name)
+		if kerr != nil {
+			return "", fmt.Errorf("error in configuration file %s: %s", path, kerr)
+		}
+		return key.Value(), nil
+	}
+
+	cfg.UpdateBundle, err = getKey(swupd, "BUNDLE")
+	if err != nil {
+		return nil, err
+	}
+	cfg.ContentURL, err = getKey(swupd, "CONTENTURL")
+	if err != nil {
+		return nil, err
+	}
+	cfg.VersionURL, err = getKey(swupd, "VERSIONURL")
+	if err != nil {
+		return nil, err
+	}
+
+	return cfg, nil
+}
+
+func (b *Builder) buildBundleChroots(set *bundleSet) error {
+	var err error
+	stateDir := b.Statedir
+	chrootDir := filepath.Join(stateDir, "image")
+
+	// TODO: Remove remaining references to outputDir. Let "build update" take care of
+	// bootstraping or cleaning up.
+	outputDir := filepath.Join(stateDir, "www")
+
+	if stateDir == "" || stateDir == "/" {
+		return errors.Errorf("invalid state dir: %s", stateDir)
+	}
+
+	if _, ok := set.Bundles["os-core"]; !ok {
+		return fmt.Errorf("os-core bundle not found")
+	}
+
+	// Bootstrap the directories.
+	err = os.MkdirAll(filepath.Join(chrootDir, "0"), 0755)
+	if err != nil {
+		return err
+	}
+	err = os.MkdirAll(filepath.Join(outputDir, "0"), 0755)
+	if err != nil {
+		return err
+	}
+	// TODO: Move this outside this function. chroot building doesn't use LAST_VER.
```

Comment:
Why not move it out now and call it only after building finishes?

Reply:
Moved outside, but added a TODO to consider moving it to the bundle update logic.

Not doing that other move now to contain the amount of change (don't want to touch "build update" in this patch).